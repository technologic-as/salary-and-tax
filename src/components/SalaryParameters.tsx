import FormControlLabel from '@material-ui/core/FormControlLabel';
import Radio from '@material-ui/core/Radio';
import React from 'react';
import { defineMessages, useIntl } from 'react-intl';
import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';
import { inputHasChanged } from '../actions';
import { State } from '../reducers';
import { Button, Checkbox, FormGroup, InputWithLabel, RadioGroup, Section } from './Ui';
import { Dispatch } from 'redux';
import { SupportedLocale } from '../constants';

const messages = defineMessages({
  header: { id: 'salary.parameters.header', defaultMessage: 'Parameters' },
  turnover: { id: 'salary.parameters.turnover', defaultMessage: 'Turnover' },
  hoursPerYear: {
    id: 'salary.parameters.hours.per.year',
    defaultMessage: 'Hours per year',
  },
  hourRate: { id: 'salary.parameters.hour.rate', defaultMessage: 'Hour rate' },
  subcontractor: {
    id: 'salary.parameters.subcontractor',
    defaultMessage: 'Subcontractor',
  },
  theCut: {
    id: 'salary.parameters.subcontractor.cut',
    defaultMessage: 'Cut (%)',
  },
  employerFee: {
    id: 'salary.parameters.employer.fee',
    defaultMessage: 'Employer Fee',
  },
  percentage: { id: 'salary.parameters.percentage', defaultMessage: '(%)' },
  includeInCalculation: {
    id: 'salary.parameters.include.in.calculation',
    defaultMessage: 'Include in calculation',
  },
  vacationSavings: {
    id: 'salary.parameters.vacation.savings',
    defaultMessage: 'Vacation savings',
  },
  pension: { id: 'salary.parameters.pension', defaultMessage: 'Pension' },
  pensionStep1: {
    id: 'salary.parameters.pension.step1',
    defaultMessage: '(%) (1G-7.1G)',
  },
  pensionStep2: {
    id: 'salary.parameters.pension.step2',
    defaultMessage: '(%) (7.1G-12G)',
  },
  surplus: { id: 'salary.parameters.surplus', defaultMessage: 'Surplus' },
  surplusAmount: {
    id: 'salary.parameters.surplus.amount',
    defaultMessage: 'Amount',
  },
  locale: { id: 'salary.parameters.locale', defaultMessage: 'Language' },
  submit: {
    id: 'salary.parameters.submit',
    defaultMessage: 'Calculate salary',
  },
});

type Props = {
  locales: { label: string; value: string }[];
  handleSubmit: () => void;
  submitting: boolean;
};
export const SalaryParametersComponent = ({ handleSubmit, submitting, locales }: Props): JSX.Element => {
  const { formatMessage } = useIntl();
  return (
    <Section header={formatMessage(messages.header)} expanded>
      <form onSubmit={handleSubmit} style={{ display: 'flex', alignItems: 'flex-start' }}>
        <FormGroup header={formatMessage(messages.turnover)}>
          <InputWithLabel name="hoursPerYear" label={formatMessage(messages.hoursPerYear)} />
          <InputWithLabel name="hourRate" label={formatMessage(messages.hourRate)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.subcontractor)}>
          <InputWithLabel name="cut" label={formatMessage(messages.theCut)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.employerFee)}>
          <Checkbox name="employerFee.include" label={formatMessage(messages.includeInCalculation)} />
          <InputWithLabel name="employerFee.rate" label={formatMessage(messages.percentage)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.vacationSavings)}>
          <Checkbox name="vacationSavings.include" label={formatMessage(messages.includeInCalculation)} />
          <InputWithLabel name="vacationSavings.rate" label={formatMessage(messages.percentage)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.pension)}>
          <Checkbox name="pension.include" label={formatMessage(messages.includeInCalculation)} />
          <InputWithLabel name="pension.step1.rate" label={formatMessage(messages.pensionStep1)} />
          <InputWithLabel name="pension.step2.rate" label={formatMessage(messages.pensionStep2)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.surplus)}>
          <Checkbox name="surplus.include" label={formatMessage(messages.includeInCalculation)} />
          <InputWithLabel name="surplus.amount" label={formatMessage(messages.surplusAmount)} />
        </FormGroup>

        <FormGroup header={formatMessage(messages.locale)}>
          <RadioGroup name="locale">
            {locales.map((locale) => (
              <FormControlLabel value={locale.value} control={<Radio />} label={locale.label} key={locale.value} />
            ))}
          </RadioGroup>
        </FormGroup>

        <Button type="submit" disabled={submitting}>
          {formatMessage(messages.submit)}
        </Button>
      </form>
    </Section>
  );
};

export const form = 'parameters';
// @ts-ignore
export const SalaryParametersForm = reduxForm({ form })(SalaryParametersComponent);

const mapStateToProps = ({ parameters, intl: { locales } }: State) => ({
  initialValues: parameters,
  destroyOnUnmount: false,
  locales,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onSubmit: (values: { locale?: SupportedLocale }) => inputHasChanged(values)(dispatch),
});

export const SalaryParameters = connect(mapStateToProps, mapDispatchToProps)(SalaryParametersForm);
export default SalaryParameters;
