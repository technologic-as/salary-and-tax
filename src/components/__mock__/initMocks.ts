const mockFormatMessage = ({ defaultMessage }: { defaultMessage: string }) => defaultMessage;
const mockFormatNumber = (amount: number) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount);

const mockIntl = () => ({
  formatMessage: mockFormatMessage,
  formatDate: jest.fn,
  formatTime: jest.fn,
  formatRelative: jest.fn,
  formatNumber: mockFormatNumber,
  formatPlural: jest.fn,
  formatHTMLMessage: jest.fn,
  now: jest.fn,
});

jest.mock('react-intl', () => ({
  useIntl: mockIntl,
  defineMessages: (msg: unknown) => msg,
}));

jest.mock('react-intl-redux', () => ({
  updateIntl: (a: unknown) => a,
}));

jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/FormGroup', () => 'FormGroup');
jest.mock('@material-ui/core/FormControlLabel', () => 'FormControlLabel');
jest.mock('@material-ui/core/Radio', () => 'Radio');

export {};
