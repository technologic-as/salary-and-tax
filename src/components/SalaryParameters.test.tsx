import mockComponent from 'identity-obj-proxy';
import React from 'react';
import renderer from 'react-test-renderer';
import './__mock__/initMocks';
import { SalaryParametersComponent } from './SalaryParameters';

const handleSubmit = () => jest.fn();

jest.mock('./Ui', () => mockComponent);

describe('SalaryParametersComponent', () => {
  it('should render', () => {
    const tree = renderer
      .create(
        <SalaryParametersComponent
          handleSubmit={handleSubmit}
          submitting={false}
          locales={[
            { value: 'en', label: 'English' },
            { value: 'nb', label: 'Norsk' },
          ]}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
