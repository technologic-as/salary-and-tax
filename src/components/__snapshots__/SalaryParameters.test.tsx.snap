// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalaryParametersComponent should render 1`] = `
<Section
  expanded={true}
  header="Parameters"
>
  <form
    onSubmit={[Function]}
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
      }
    }
  >
    <FormGroup
      header="Turnover"
    >
      <InputWithLabel
        label="Hours per year"
        name="hoursPerYear"
      />
      <InputWithLabel
        label="Hour rate"
        name="hourRate"
      />
    </FormGroup>
    <FormGroup
      header="Subcontractor"
    >
      <InputWithLabel
        label="Cut (%)"
        name="cut"
      />
    </FormGroup>
    <FormGroup
      header="Employer Fee"
    >
      <Checkbox
        label="Include in calculation"
        name="employerFee.include"
      />
      <InputWithLabel
        label="(%)"
        name="employerFee.rate"
      />
    </FormGroup>
    <FormGroup
      header="Vacation savings"
    >
      <Checkbox
        label="Include in calculation"
        name="vacationSavings.include"
      />
      <InputWithLabel
        label="(%)"
        name="vacationSavings.rate"
      />
    </FormGroup>
    <FormGroup
      header="Pension"
    >
      <Checkbox
        label="Include in calculation"
        name="pension.include"
      />
      <InputWithLabel
        label="(%) (1G-7.1G)"
        name="pension.step1.rate"
      />
      <InputWithLabel
        label="(%) (7.1G-12G)"
        name="pension.step2.rate"
      />
    </FormGroup>
    <FormGroup
      header="Surplus"
    >
      <Checkbox
        label="Include in calculation"
        name="surplus.include"
      />
      <InputWithLabel
        label="Amount"
        name="surplus.amount"
      />
    </FormGroup>
    <FormGroup
      header="Language"
    >
      <RadioGroup
        name="locale"
      >
        <FormControlLabel
          control={<Radio />}
          label="English"
          value="en"
        />
        <FormControlLabel
          control={<Radio />}
          label="Norsk"
          value="nb"
        />
      </RadioGroup>
    </FormGroup>
    <Button
      disabled={false}
      type="submit"
    >
      Calculate salary
    </Button>
  </form>
</Section>
`;
